lib/class-wp-theme-json-resolver-gutenberg.php

```diff
@@ -531,3 +531,4 @@ public static function clean_cached_data() {
 }
 
 add_action( 'switch_theme', array( 'WP_Theme_JSON_Resolver_Gutenberg', 'clean_cached_data' ) );
+add_action( 'setup_theme', array( 'WP_Theme_JSON_Resolver_Gutenberg', 'clean_cached_data' ) );
```

Comment:
I wonder if it'd be enough to only hook to `setup_theme`. In other words: is there any situation in which the `switch_theme` is fired but the `setup_theme` isn't?

The current fix works, though, and the worst-case scenario is that we read the data from the sources twice for a theme.

Reply:
I believe `setup_theme` alone could be enough. `[switch_theme](https://developer.wordpress.org/reference/hooks/switch_theme/)` is triggered when we deactivate a theme, which is not the case for the previews.

Maybe @aristath, @carolinan  or @swissspidy could confirm? 😄

Reply:
`setup_theme` is fired before the current theme is loaded when accessing WordPress (happens late in `wp-settings.php`). This only happens once per request. This is similar to the `plugins_loaded` hook.

`switch_theme()` (which fires the `switch_theme` action) can be called later on and potentially multiple times during a request, for example when publishing customizer changesets, upgrading the theme, or of course when switching themes.

For this reason, it's definitely needed to hook into `switch_theme`

Not super familiar with the Customizer, but `WP_Customize_Manager` hooks into `setup_theme` to preview different themes when accessing WordPress.

That makes sense, since the Customizer loads the WP site in an iframe and reloads that page when changing themes. So each time it's a new request, firing `setup_theme` anew, then kicking off the customizer logic.

Hooking into `setup_theme` to clean cached data seems a bit odd to do this.

In a regular request, there's no data to clear yet, and for customizer requests it just so happens that your code runs before the customizer.

It would be better to hook into the customizer logic itself.

**tl;dr:**

1. Hook into `switch_theme`
2. Hook into `start_previewing_theme`

Reply:
Thanks for the explanation, Pascal! I've implemented what you suggested.

I agree using `setup_theme` is a bit odd (cleaning the cache before the theme data is even used, in most cases). However, I wonder if that's the hook that's most relevant to what we want to do here, instead of hooking into flow-specific ones. There may be flows that are not covered yet. I'll keep an eye on this.

Reply:
I've run into a meta trac ticket that may be related to this https://meta.trac.wordpress.org/ticket/5818#comment:4 (previews for themes with theme.json not working consistently in the theme directory).